Unity-Technologies/com.unity.mobile.logcat
# Memory Window + Minor Fixes

**Minor Fixes:**
* Fixed cases where executing async operation which throw exception inside, stops the dispatcher and half of the logcat simply stops working, until Editor is restarted or domain reload occurs, affected areas:
   * Connect to IP operation
   * Device Search operation
* Added bunch of test for dispatcher
* Added some more diagnostics when Internal build is enabled, it will show now how many async operations are in queue, and how many are executed
* Android Logcat state is now saved in ProjectSettings
* Android Logcat state is now correctly saved/restore during domain reloads , this fixes https://github.com/Unity-Technologies/com.unity.mobile.logcat/issues/71
* Added Capture button in Capture Screenshot window
* Capture Screen doesn't hang main thread now
* Fixed https://github.com/Unity-Technologies/com.unity.mobile.logcat/issues/72, if the error will happen, it won't interrupt the message parsing
* Added Tools menu, various tools like Open Terminal, StackraceUtility, Capture Screen, Memory Window were placed there
* Added Memory Tests project
* Clean package list in Select Package menu, only max 5 Exited package will be shown
* Optimized UpdateDebuggablePackages list, since it was querying ip from the same package several times
* Internal log has a Clear button

**Memory Window**
![MemoryWindow](https://user-images.githubusercontent.com/568752/78994021-d5517000-7b47-11ea-8f6d-fc28b8b231a7.png)
* Checkout video in https://drive.google.com/drive/u/2/folders/1VlEfDMn7tIdMiyThDzAQHr_jZkm0vta0
* Add an ability to view various memory types of a package
* It has different memory groups, please read docs about them
* Compare to Unity's Profiler which only shows memory allocated by internal Unity memory managers, Android Logcat Memory Window will show Total memory

How to test?
* Verify docs, by opening com.unity.mobile.logcat\TestProjects\SampleProject1, go to Package Manager, select Android Logcat, click Generate Documentation
* How to check Memory Window, there's a helper project TestProjects\MemoryViewer, it's able allocate/deallocate/dirty memory
* Build to Android with Development checkbox On (optional), run the app
* Hook the Android Logcat and select the app package
* Memory Window will appear, and start request memory every 500 ms (can be changed in settings)
* As a default, Heap Alloc is selected, it will show total native/java memory allocated which is not necesserily dirty
* On Android, click Allocate Native/Java Memory, Deallocate. See how the chart reflects these changes
* To check dirty memory, first select Proportional Set Size group, assuming you didn't deallocate the memory, click dirty, the memory should be transferred to Android's RAM and thus appear in Proportional Set Size graph
* You could also check what UNity's profiler is showing for memory, it should show much less.
* You could also compare this to ANdroid's studio profiler, which shows Proportional Set Size, which should match what Android Logcat is showing. 
* You could try plugging out USB cable, and see if this affect Memory Viewer in a bad way

What else to test?
* Do this routine 6 times, launch app, select it in logcat, quit app (repeat 6 times), the app with different process id, will fill select package list, but after 5 exited entries, it should no longer grow
* Check that your added tags are still there after exiting editor and launching it again
* Check Capture Screen's Capture button
